### PR TITLE
Fix sqlalchemy 2 compatibility

### PR DESCRIPTION
**Pull request by tonykploomber**
## Describe your changes

- Upgrade to sqlalchemy 2 
- Fix test cases

## Issue number

Closes https://github.com/ploomber/jupysql/issues/234

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#lintingformatting)
- [ ] Added [tests](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#testing) (when necessary).
- [ ] Added documentation in the [docstring](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#documenting-changes-and-new-features) and [changelog](https://github.com/ploomber/contributing/blob/main/CONTRIBUTING.md#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--339.org.readthedocs.build/en/339/

### CODE REVIEW

**Comment by edublancas** [comment]
don't forget to add the version checks

https://github.com/ploomber/jupysql/blob/f299432fb767e7da9a161317b70f9870f18a0f0e/.github/workflows/ci.yaml#L45
https://github.com/ploomber/jupysql/blob/f299432fb767e7da9a161317b70f9870f18a0f0e/.github/workflows/ci.yaml#L80

because this caused us trouble last time: the pip resolver might end up installing sqlalchemy 1.x so we must ensure that we're actually testing with both sqlalchemy 1.x and 2.x

**Comment by tonykploomber** [comment]
@edublancas 
I found 2 issues during the migration process in sqlalchemy.

1. In duckdb, seems our `_commit` implementation in the `sql.run` will cause the issue: User's Query Result will be impacted by our `_commit` method while execution:

Experimental: duckdb - Fail 
```python3
# Please upgrade to sqlalchemy2 to test
from sqlalchemy import create_engine, MetaData, text
import pandas as pd
eng = create_engine("duckdb://").connect()

# Test data
d = {'col1': [1, 2], 'col2': [3, 4]}
df = pd.DataFrame(data=d)
df.to_sql("test", con=eng, if_exists="replace", index=False)

result =eng.execute(text("SELECT * FROM test LIMIT 3")) # Original User SQL clause
eng.execute(text("COMMIT")) # We have _commit method will trigger this

result.fetchall() # Seems previous saved result from original SQL clause will be impacted by "COMMIT" execution in duckDB

# Output: []
```

Other Database - Normal 
```python3
# Please upgrade to sqlalchemy2 to test
from sqlalchemy import create_engine, MetaData, text
import pandas as pd
eng = create_engine("duckdb://").connect()

# Test data
d = {'col1': [1, 2], 'col2': [3, 4]}
df = pd.DataFrame(data=d)
df.to_sql("test", con=eng, if_exists="replace", index=False)

result =eng.execute(text("SELECT * FROM test LIMIT 3")) # Original User SQL clause
eng.execute(text("COMMIT")) # We have _commit method will trigger this

result.fetchall() # Seems previous saved result from original SQL clause will not be impacted by "COMMIT" execution in duckDB

# Output: []
```

I am not sure if this is the issue between duckdb and sqlalchemy seems our other databases work fine

**Comment by tonykploomber** [comment]
Another issue is in Python 3.7, user will not able to use `_persist_dataframe` feature if we migrate to sqlalchemy 2

Context: We passed all Python version's unit tests but only Python 3.7 won't pass, turned out the Python 3.7 only can use the Pandas 1.3.5 below versions, which that version of pandas implementation of that isn't support the sqlalchemy2

We use `pandas.to_sql` to have this feature, they the way they initial MetaData Class from sqlalchemy isn't matching to version 2

Pandas 1.3.5, [Source code](https://github.com/pandas-dev/pandas/blob/1.3.x/pandas/io/sql.py#L1410) 
```
    def __init__(self, engine, schema: str | None = None, meta=None):
        self.connectable = engine
        if not meta:
            from sqlalchemy.schema import MetaData

            # This way only applicable to 
            meta = MetaData(self.connectable, schema=schema)

        self.meta = meta
```


In sqlalchemy1, [the Metadata Class](https://docs.sqlalchemy.org/en/14/core/metadata.html#sqlalchemy.schema.MetaData) can accept the engine as argument, that's panda's 1.3.5's implementation

In sqlalchemy2, [the Metadata Class](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.MetaData.__init__) doesn't accept engine as argument anymore

The newest version of `pandas 1.5.x` change the implementation of MetaData (how sqlalchemy2 defined), but this version doesn't support `python3.7`

## TL;DR; 

Our _persist_dataframe will use pandas `to_sql` feature, but the impelmentation of `to_sql` on that pandas 1.3.5 did not compatible to the API in sqlalchemy2 in Python3.7

## Plans

1. In the PR we can still go ahead to upgrade to sqlalchemy2
2. For `_persist_dataframe` feature, let's just add the warning message python 3.7 to tell users this feature can't be supported anymore

Python3.7 will be deprecated in 2 months, I think it's reasonable move for now

**Comment by edublancas** [comment]
yeah, makes sense. we can detect if user has py 3.7 and sqlalchemy 2. in such case we can tell them they can either downgrade sqlalchemy or upgrade python. python 3.7 is already giving trouble for many libraries so we already deprecated it in sklearn-evaluation so it makes sense to deprecate it soon.

**Comment by edublancas** [comment]
I just realized that py 3.7 will be deprecated this summer, so I'd even go as far as dropping support for it now: https://endoflife.date/python

**Comment by tonykploomber** [comment]
### Python 3.7
Agree with dropping 3.7, but we should also add some note to `--persist` doc that Python 3.7 might not be able to use since we now aware of this

---

### Commit issue

I fix by changing the way to autocommit in https://github.com/ploomber/jupysql/pull/339/commits/0076148bf8170d8b7ced5d2172b4d02f306c18cc

However, that raises two issues:
1. Failing in the sqlite plugin in duckdb, now after the installation of plugin user needs to commit it manually

Originally
```
%sql INSTALL 'sqlite_scanner';
%sql LOAD 'sqlite_scanner';
%sql CALL sqlite_attach('test_db.db', overwrite=true);
```

Now, the user needs to commit manually after the installation
```
%sql INSTALL 'sqlite_scanner';
%sql LOAD 'sqlite_scanner';
%sql commit;
%sql CALL sqlite_attach('test_db.db', overwrite=true);
```

So I have to update the test_plot [in here](https://github.com/ploomber/jupysql/pull/339/commits/c5e65e6d794d252e3b00949a2d8de7d9a96c6d0b), probably this is more like the issue between duckdb and sqlalchemy.

2. Failing in test case `test_inspect.py::test_nonexistent_table `, it was success on my local but not on CI, wasn't able to reproduce the issue

CI Fail link: https://github.com/ploomber/jupysql/actions/runs/4575332985/jobs/8078038888

Conclusion
For 1. Should we raise the issue in duckdb or duckdb-engine?
For 2. Can you help me to take a look? Are you able to reproduce the issue in your local?

**Comment by edublancas** [comment]
for DuckDB: what I understand is that the autocommit feature (which is turned on by default) is causing SQL queries to return empty results.

I think let's open an issue in duckdb-engine with a reproducible example and ask if this expected (I'm guessing this is a problem with duckdb-engine rather than duckdb). Then, we decide.

**Comment by tonykploomber** [comment]
@edublancas 
Yes, that's why I was having a difficult time, so I change the way we implement autocommit in https://github.com/ploomber/jupysql/commit/0076148bf8170d8b7ced5d2172b4d02f306c18cc.
Then the empty results issue is gone.

Since we are not executing "`commit`" anymore, then it's causing [this issue](https://github.com/Mause/duckdb_engine/issues/593) (which raised in duckdb-engine)

Let's check it's this is the plugin issue between duckdb-engine & sqlalchemy2. Ideally if they can solve this issue then we are good.

I think returning empty result is a bigger issue than the plugin issue. Let's see then

**Comment by tonykploomber** [comment]
Should be ready for review

**Comment by tonykploomber** [comment]
1. Added CHANGEDLOG
2. Update doc - https://jupysql--339.org.readthedocs.build/en/339/integrations/pandas.html
Screenshot 2023-04-05 at 9 30 08 AM

**Comment by flaviomartins** [comment]
> @idomic @tonykploomber: I'm finalizing tests in the master branch to release 0.7, let's _not merge_ this until we release it

@edublancas I think that means supporting two versions going forward?
- 0.7 with sqlalchemy 1.x
- 0.8 with sqlalchemy 2.x

**Comment by edublancas** [comment]
@idomic, @tonykploomber: I've created the tag for the 0.7 release. we can merge this (once it's ready, I see there are some merge conflicts)

> @edublancas I think releasing 0.7 without it could mean having to support two versions

JupySQL 0.7 only supports SQLAlchemy 1.x, but beginning in the next release, we'll support both SQLAlchemy 1.x and 2.x, and keep the support for both for a few major releases.

> Disclosure: I've been pip installing a merge of this PR in my jupyter docker image to use sqlalchemy2. What would be the timeframes for the releases?

I think this is almost ready, we're aiming to have this on PyPI next week.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -32,10 +32,13 @@ jobs:
...
+          pip install sqlalchemy -U
 
       - name: Test with pytest
         run: |
+          # ensure we're running sqlalchemy 2
+          python -c "import sqlalchemy; assert int(sqlalchemy.__version__.split('.')[0]) == 2"
+          # run unit tests
           pytest --durations-min=5 --ignore=src/tests/integration
 
   test-sqlalchemy-v1:
```
[the remark this answers is not in the document]
Comment (tonykploomber) [reply .github/workflows/ci.yaml:44]: @edublancas 
Do you know the clear context of this to answer Ido? Since I was referring to the previous PR to do this

For my understanding:
This is making sure the CI container uses sqlalchemy 2, if the assert condition is false, the test case is early failed

**Comment by tonykploomber** [comment]
@idomic Fix, thanks for the detail review!

File: .github/workflows/ci.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (edublancas) [reply .github/workflows/ci.yaml:44]: we broke a previous release because the pip resolver was installing sqlalchemy 1.x even though we were doing `pip install sqlalchemy`; this is a sanity check to ensure we're actually testing against sqlalchemy 2.x

this will add some CI extra time since we're adding new jobs. but we're testing sqlalchemy 1.x only with Python 3.11. and since github runs the build jobs in parallel, it won't have much of an effect. We can decide later if we expand or reduce this. I think for now, it's fine
